internal/speakerlist/speakerlist.go

```diff
@@ -92,16 +108,155 @@ func (sl *SpeakerList) Start(client *k8s.Client) error {
+			iplist, err := sl.mlSpeakers()
+			if err != nil {
+				continue
+			}
+
+			sl.mlMux.Lock()
+			sl.mlSpeakersIPs = iplist
+			sl.mlMux.Unlock()
+		}
+	}
+}
 
+func (sl *SpeakerList) mlSpeakers() ([]string, error) {
+	// This call blocks until we get a response of the API server.
+	// In the client-go version we are using, there is no way to use a
+	// context or timeout.
+	// TODO: When updating client-go, this code can be simplified by using a
+	// context in the following way:
+	// If we use a context, we can specify a timeout and there is no need to
+	// have the speakersIPsLoop() running in the background. This can just
+	// be called from mlJoin() with a context (timeout) and if doesn't
+	// answer, use sl.mlSpeakersIPs.  If it doesn't time out, just update
+	// sl.mlSpeakersIPs with the answer and continue to do the join.
 	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
+	if err != nil {
+		sl.l.Log("op", "Member Detection", "error", err, "msg", "failed to get PodsIPs")
+		return nil, err
+	}
+
+	return iplist, nil
+}
+
+func (sl *SpeakerList) mlJoinLoop() {
+	iplist, err := sl.mlSpeakers()
 	if err != nil {
 		sl.l.Log("op", "startup", "error", err, "msg", "failed to get PodsIPs")
-		return err
+		iplist = nil
 	}
-	n, err := sl.mList.Join(iplist)
-	sl.l.Log("op", "startup", "msg", "Memberlist join", "nb joigned", n, "error", err)
 
-	return nil
+	// Initialize sl.mlSpeakersIPs.
+	sl.mlMux.Lock()
+	sl.mlSpeakersIPs = iplist
+	sl.mlMux.Unlock()
+
+	// Update mlSpeakersIPs in the background.
+	go sl.speakersIPsLoop()
+
+	// Every one minute, try to rejoin.
+	// This will join nodes that leave the cluster only for a few seconds every 1
+	// minute, while discovering new IPs (updated by sl.speakersIPsLoop()) might
+	// take a while longer.
+	ticker := time.NewTicker(1 * time.Minute)
+
+	for {
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			sl.mlJoin()
+		case <-sl.mlJoinCh:
+			sl.mlJoin()
+		}
+	}
+}
+
+func (sl *SpeakerList) members() map[string]struct{} {
+	members := make(map[string]struct{})
+
+	for _, node := range sl.mList.Members() {
+		ip := node.Addr.String()
+		members[ip] = struct{}{}
+	}
+
+	return members
+}
+
+// mlJoin joins this memberlist cluster with other speakers pods that are not
+// currently member of this cluster. In other words, it does a memberlist.Join()
+// with the IPs in mlSpeakersIPs that are not members of the cluster.
+func (sl *SpeakerList) mlJoin() (nr int, err error) {
+	// Calculate IPs to join, IPs not currently members of the memberlist cluster.
+	var joinIPs []string
+
+	members := sl.members()
+
+	sl.mlMux.Lock()
+	defer sl.mlMux.Unlock()
+
+	for _, ip := range sl.mlSpeakersIPs {
+		// If an ip is not member of the cluster, add it to joinIPs.
+		if _, isMember := members[ip]; !isMember {
+			joinIPs = append(joinIPs, ip)
+		}
+	}
+
+	if len(joinIPs) == 0 {
+		// Not logging here to not spam the logs
+		return 0, nil
+	}
+
+	for i := 0; i < 2; i++ {
```

Comment:
I'm not sure we should retry multiple time
To join the memberList you just need to join 1 of the members and you will rejoin the whole network.
Partial join means one node is really down I think

Reply:
Down or not ready.

I'm fine without a retry, because we will retry in a minute anyways if something fails. Just that the time to join a node might be longer. Are you okay with that too? :)

Reply:
I prefer to have this fix simple (but correct) as we will likely remove most of it in the next minor release anyway

Reply:
Cool, removed!